### ISSUE TITLE
Offline version of the simulator

### ISSUE THREAD

**Issue by urish** [wokwi/wokwi-features#204]
A version of the simulator you can use without an internet connection. Either by making the online version a PWA (progressive web app) that is available offline, or by providing a downloadable version of the application. In any case, this will require installing an Arduino compiler on the local machine (in order to compile the code).

Note that the Offline version is very likely to be a paid feature. Anyone who votes for this feature (before the first version is released) will get this feature, including one year of updates. [Wokwi club members](https://wokwi.com/club) will also get access to the offline version.

**Comment by MykolaKyrylenko** [comment]
A version that can be integrated into an LMS (e.g. Moodle) will be great.

**Comment by urish** [comment]
> A version that can be integrated into an LMS (e.g. Moodle) will be great.

You mean, a way to embed simulation projects into Moodle? Or do you have in mind a different kind of integration?

**Comment by MykolaKyrylenko** [comment]
What what I have seen, there are several ways to add to Moodle.
- With Falstad, I managed to compile the code and upload the resultant website as a file onto Moodle. Details here: https://help.lafayette.edu/importing-a-website-into-moodle/
- I have not done this, but another option could be H5P, which is HTML5 content

Note: the Falstad implementation includes AVR8js which works well on our Moodle

**Comment by urish** [comment]
Interesting! I'm not very familiar with Moodle, so this is useful for me. Mind opening a different issue for the possible Moodle integration?

**Comment by homevk15** [comment]
Hi. I'm very interested in an offline version of the simulator. Please, tell me how I can get it. Will I get source code too? Will it be compiled in the same website as wokwi.com? Will it have Arduino and Raspberry Pi Pico?  Is it allowed to make it available online and embed it on my website for others for free?

**Comment by gauravacad** [comment]
cannot we use limited data for IoT gateway http server. always having club feature is not possible

**Comment by sasdallas** [comment]
@gauravacad Assuming there is no solution, is there a way to mark this as unaddable or so? As disappointing as it is, if it actually cannot happen, most members shouldn't waste their votepowers on it.

**Comment by urish** [comment]
> As disappointing as it is, if it actually cannot happen, most members shouldn't waste their votepowers on it.

Are you asking about Offline version of the simulator? It's in development. It'll be available as a feature in Wokwi for VS Code, and perhaps also as a totally standalone version after, depending on demand.

**Comment by urish** [comment]
Wokwi for VS Code now has initial support for offline operation, as a paid add-on.  You can purchase it when you activate your Wokwi for VS Code license, from this page: https://wokwi.com/license.

I would love to hear your feedback - does it answer your needs? Is anything missing?

**Comment by sasdallas** [comment]
@urish While I understand the fact that bills need to be paid, $97 a year seems quite pricy, especially for those with big projects who just want to compile it faster.
I haven't tried it out yet but maybe adding a monthly payment option or a free trial might be helpful! This definitely seems like something bigger than Wokwi Club, so it does make sense to have it as an addon.

Thank you for your work!

**Comment by urish** [comment]
Thanks for the feedback @sasdallas! I'll look into adding a monthly payment option.

**Comment by Timmmm** [comment]
I think it would be worth clarifying somewhere whether Wokwi sends code to your servers or not if you *don't* buy the offline feature. Your privacy policy doesn't say anything about this - is the online requirement for the free version of the VSCode extension just a "check we're online" type thing or is it actually sending stuff to servers?

**Comment by urish** [comment]
> I think it would be worth clarifying somewhere whether Wokwi sends code to your servers or not if you don't buy the offline feature. Your privacy policy doesn't say anything about this - is the online requirement for the free version of the VSCode extension just a "check we're online" type thing or is it actually sending stuff to servers?

Thanks for the feedback!

Wokwi for VS Code currently does not send your code or your compiled firmware binaries to the server. It does send some pseudo anonymized telemetry which includes information about your VS Code version, workspace type (detected by checking if files such as platformio.ini, CMakeFiles.txt, or Cargo.toml) exists, OS version, etc, unless VS Code tells it [not to send any telemetry](https://code.visualstudio.com/api/extension-guides/telemetry#without-the-telemetry-module).

In case we learn people are using the free version of Wokwi for VS Code to work on non open source projects, we may decide to introduce a mechanism that actually verifies if the project is open source. For this reason, the privacy policy leaves this question open.

I hope this helps!

**Comment by Timmmm** [comment]
Very helpful thanks! Though I'm not sure you can "leave questions open" in your privacy policy. If you don't say you are using some data then you aren't allowed to use it. However I think you can just update the privacy policy if/when you want to change it.

**Comment by selmen2004** [comment]
@urish I have complied an offline version using only ESP32 / micropython for educational use only ( we teach using wokwi in our secondary schools in Tunisia and some schools have internet issues )
is this ok  ?

**Comment by urish** [comment]
> @urish I have complied an offline version using only ESP32 / micropython for educational use only ( we teach using wokwi in our secondary schools in Tunisia and some schools have internet issues )

Would love to see what you compiled - how can I check it out?

**Comment by selmen2004** [comment]
> > @urish I have complied an offline version using only ESP32 / micropython for educational use only ( we teach using wokwi in our secondary schools in Tunisia and some schools have internet issues )
> 
> Would love to see what you compiled - how can I check it out?

Sure , here  : https://selmen.tn/2023/04/28/wokwi-esp32-micropython-simulator-offline/

**Comment by saturnquack** [comment]
Any update on an offline version of the non-vs code one? From what I'm reading, its a maybe?

**Comment by urish** [comment]
> Any update on an offline version of the non-vs code one? From what I'm reading, its a maybe?

Nothing official at the moment :)

**Comment by the-mentor** [comment]
I would also like an offline version of the non VSCode one

I want to use it in my open source  ESPHome Simulator if possible 
https://github.com/the-mentor/ESPHome-Device-Sim/

Keep up the amazing work !!

**Comment by kyvaith** [comment]
Does the free version of the Wokwi extension for VSCode do the simulation/emulation in the cloud, while the offline version does it locally? I'm trying to find a way to speed up the emulation, because now trying to emulate LVGL stuff gives me 1-5 FPS compared to 60 FPS on the real ESP32.

**Comment by urish** [comment]
> I'm trying to find a way to speed up the emulation, because now trying to emulate LVGL stuff gives me 1-5 FPS compared to 60 FPS on the real ESP32

Which display are you using and which specific ESP32 chip? Speeding up LVGL simulation is a real challenge.

**Comment by kyvaith** [comment]
I'm using the only avaliable touchscreen display in Wokwi. The Chip is ESP32-S3.

**Comment by urish** [comment]
Yeah, the S3 is especially challenging, since we have to simulate two cores. If you can recreate a minimal reproduction of this and open a new issue, we can take a look if there are any quick wins (but I wouldn't expect more than x2-x3 speed up compared to the current simulation speed, without a significant amount of optimization work)

**Comment by kyvaith** [comment]
I'm using ESPhome in VSCode, so probably no good example to bo pasted in the online version of Wokwi. I can provide you compiled firmware or you can compile it from my code https://github.com/kyvaith/atmosfera-echo-hub

**Comment by urish** [comment]
Thanks, can you please open a separate issue to track this? also, please explain how to compile the code in the repo (I see some files there, but there are no sources, as I assume there's some build process with esphome)

**Comment by kyvaith** [comment]
Ok, but the main question is if the Wokwi emulates ESP32 localy or in cloud. If it's localy maybe I can just use machine with more powerfull CPU? Regarding the compilation proces - just use the instructions from my repo.

**Comment by urish** [comment]
Locally

**Comment by brentru** [comment]
@urish  Could you please clarify if the Offline version of the simulator is only for VSCode, or can we use it for running Wokwi-cli test scenarios offline (instanciated from a terminal)?

**Comment by urish** [comment]
@brentru there's an offline version of the wokwi-cli server that can run in docker, so it can be used offline. Would love to hear more about your use case

**Comment by sasdallas** [comment]
> Payed? Greedy as shit of something to do tbh

Not at all. It is not easy to create a simulator that can handle this level of firmware emulation and having devices act as they should in simulation. Emulating an Arduino is hard work, emulating an arduino that has I/O busses and communicates with other peripherals is even more work.

While I agree with you that the current price is _way_ out of proportion (at least when I checked), being paid isn't a greedy move.

Edit: The price is actually reasonable, I misremembered it as $97/month and not a year. If @urish adds monthly options I would gladly grab a license.

**Comment by salvalie** [comment]
@MicrocadeGuy It's a proprietary software, you are in no position to discuss their business model. Just dont buy it. Is it too hard to understand? I am looking into foss alternatives, but those aren't even close to this software.

**Comment by robherc** [comment]
Is there any way we could add the ability for those of us who own servers & or cloud vservers to install a wokwi "slave" on our public servers to "donate compile time" to wokwi to help with the compile servers' loads, in exchange for our own builds being able to be priority-executed on our own server?

in essence:
I install the wokwi slave app on my public server; set max concurrent builds that I'm willing to donate of server cpu-time; then my server loadshares some of the public/dedicated builds for wokwi, and any time I start a wokwi build, it automatically compiles on my server (bypassing the preset max concurrent builds, or just always "next in line" ahead of already queued builds)

**Comment by urish** [comment]
Closing as we now have offline capabilities both in Wokwi for VS Code (as part of the Pro plan) and in the Wokwi CLI.

**Comment by djedu28** [comment]
> Closing as we now have offline capabilities both in Wokwi for VS Code (as part of the Pro plan) and in the Wokwi CLI.

I'm really happy with this news. Thank you, Urish, for your dedication to this project. It reflects your passion for development, the maker community, and your drive to do good and help those who don’t have the hardware to let their electronic creativity take flight.


auto-translated.
<details><summary>Original text</summary>
Fico muito feliz com essa notícia. Obrigado, Urish, pela dedicação neste projeto. Reflete a paixão que tem pelo desenvolvimento, pela comunidade maker e a motivação de fazer o bem e ajudar aqueles que não possuem hardware para dar asas a criatividade eletrônica. 
</details>
